src/ILVerify/src/ILImporter.Verify.cs

```diff
@@ -1245,26 +1246,50 @@ void StartImportingBasicBlock(BasicBlock basicBlock)
                 }
                 else
                 {
+                    Debug.Fail("Block marked as filter / handler start but no filter / handler index set.");
                     return;
                 }
 
-                if (r.ILRegion.Kind == ILExceptionRegionKind.Filter)
-                {
-                    basicBlock.EntryStack = new StackValue[] { StackValue.CreateObjRef(GetWellKnownType(WellKnownType.Object)) };
-                }
-                else
-                if (r.ILRegion.Kind == ILExceptionRegionKind.Catch)
+                if (r.ILRegion.Kind == ILExceptionRegionKind.Filter || r.ILRegion.Kind == ILExceptionRegionKind.Catch)
                 {
-                    basicBlock.EntryStack = new StackValue[] { StackValue.CreateObjRef(ResolveTypeToken(r.ILRegion.ClassToken)) };
+                    // stack must uninit or 1 (exception object)
+                    Check(basicBlock.EntryStack == null || basicBlock.EntryStack.Length == 1, VerifierError.FilterOrCatchUnexpectedStack);
+
+                    if (basicBlock.EntryStack == null)
+                        basicBlock.EntryStack = new StackValue[1];
+
+                    if (r.ILRegion.Kind == ILExceptionRegionKind.Filter)
+                    {
+                        basicBlock.EntryStack[0] = StackValue.CreateObjRef(GetWellKnownType(WellKnownType.Object));
+                    }
+                    else
+                    if (r.ILRegion.Kind == ILExceptionRegionKind.Catch)
+                    {
+                        var exceptionType = ResolveTypeToken(r.ILRegion.ClassToken);
+                        Check(!exceptionType.IsByRef, VerifierError.CatchByRef);
+                        basicBlock.EntryStack[0] = StackValue.CreateObjRef(exceptionType);
+                    }
                 }
                 else
                 {
-                    basicBlock.EntryStack = s_emptyStack;
+                    // stack must be uninit or empty
+                    Check(basicBlock.EntryStack == null || basicBlock.EntryStack.Length == 0, VerifierError.FinOrFaultNonEmptyStack);
+                    if (basicBlock.EntryStack == null)
+                        basicBlock.EntryStack = s_emptyStack;
                 }
             }
 
             if (basicBlock.EntryStack?.Length > 0)
             {
+                if (!basicBlock.TryStart && !basicBlock.HandlerStart && !basicBlock.FilterStart)
+                {
+                    // ECMA III 1.7.5 Backward Branch Constraints
+                    // if stack is not empty at beginning of this block,
+                    // there must exist a predecessor block with lower IL offset.
+                    Check(basicBlock.HasPredecessorWithLowerOffset, VerifierError.BackwardBranch);
```

Comment:
The basic block can have more than one predecessor. We can be here because we have seen the one with higher offset, but not the one with lower offset yet, and thus report error incorrectly.

Would it be better to this in InitialPass instead? This InitialPass does a single forward scan of IL that this is trying to check for.

Reply:
I actually thought that the way the current implementation imports fallthroughs would guarantee that all relevant blocks with lower offsets will be verified first.
But you are right, this is not always the case. I found another test case, which would actually produce such an incorrect error.

I will change this so the HasPredecessor flag is set inside the InitialPass and add the mentioned test case.